### ISSUE TITLE
Using it correctly?

### ISSUE THREAD

**Issue by rrwick** [thegenemyers/DASCRUBBER#7]
Hi Gene,

As I mentioned in [a previous issue](https://github.com/thegenemyers/DASCRUBBER/issues/6), I wrote [this wrapper script](https://github.com/rrwick/DASCRUBBER-wrapper) to use your tools for scrubbing long read sets, including ONT reads. I did my best to interpret your command guides, but I thought I'd ask for your input. In short, am I using your tools correctly to scrub reads?

Here are the annotated commands my wrapper would use for a bacterial read set:
```
# Creating Dazzler database
# The input file (renamed_reads.fasta) is made by the wrapper to give fake PacBio read headers to the reads.
fasta2DB reads.db renamed_reads.fasta
DBsplit -s100 reads

# Read overlap alignment with daligner
mkdir align_temp
daligner -v -Palign_temp -M80 reads reads
rm -r align_temp

# Masking repeats with REPmask
# The threshold (273) is a multiple of the base read depth (91x for this read set).
REPmask -v -c273 reads reads.reads.las

# Finding tandem repeats with datander
mkdir align_temp
datander -v -Palign_temp reads
rm -r align_temp

# Masking tandem repeats with TANmask
TANmask -v reads TAN.reads

# Read overlap alignment with daligner (with repeat masking)
mkdir align_temp
daligner -v -Palign_temp -mrep -mtan -M80 reads reads
rm -r align_temp

# Finding intrinsic quality values with DASqv
DASqv -v -c91 reads reads.reads.las

# Trimming reads and breaking chimeras with DAStrim
# The -g and -b values are taken from the DASqv suggestions.
DAStrim -v -g24 -b30 reads reads.reads.las

# Patching low quality segments with DASpatch
DASpatch -v reads reads.reads.las

# Building new database of scrubbed reads with DASedit
DASedit -v reads patched_reads

# Extracting scrubbed reads
mv renamed_reads.fasta temp.fasta
DB2fasta -vU patched_reads
mv renamed_reads.fasta scrubbed_reads.fasta
mv temp.fasta renamed_reads.fasta
```

In particular, I was wondering:
* Am I doing the repeat masking right? I first run daligner, then find repeats, then run daligner again with masks. Is that the best way?
* Alternatively, could I skip repeat masking entirely - just go from daligner to DASqv? For the purposes of scrubbing (not assembly) what are the risks of having repeat-induced local alignments?
* On a related note, I wasn't sure about the `-c` parameter for REPmask. Is it better to set it high and potentially miss repeats, or to set it low and potentially mask regions that aren't repeats?
* I'm not currently masking regions with DBdust. Do you think it's worth doing so? Are the default DBdust parameters good for most cases?

Thanks again for all your great tools!

Ryan

**Comment by thegenemyers** [comment]
Hi Ryan,

     I took a quick look at your script.  The good news is that it is 
basically correct.

     The big thing that seems to be missing is you are not taking 
advantage of and/or do not need
to split the DB in to blocks.  Basically, for really big data sets, like 
say a 60X mammal at 180Gbp
of data, you can't just say "daligner reads reads".  There isn't enough 
memory on any machine and
even if there were it would take 2-3K CPU hours (assuming it was repeat 
masked, 20K if not !).  So
the Dazzler suite has the concept of partitioning the DB into blocks 
(that's what DBsplit is for) and
then one can operate on blocks or block pairs in parallel on an HPC 
cluster,  and there are a bunch
of tools like Catrack and LAmerge for bringing the block-based results 
together into a result for the
whole data set.  For example, I generally divide a DB into 200Mbp blocks 
(-s200), and then I compare
the blocks in 16Gbp of memory with "daligner -M16 reads.3 reads.5" 
(which compares block 3 against
block 5 producing reads.3.reads.5.las).   HPC.daligner is a command that 
produces a script that does
all the block compares and then consolidates the block pairs into all 
alignments for a given block of
reads.  Similarly HPC.TANmask and HPC.REPmask produce block-based 
scripts for accomplishing
repeat masking.

     Similarly, the scrubbing commands, like DASqv and DAStrim operate 
on blocks and produce "tracks"
for each block that can then be pulled together into a single track for 
the whole DB with Catrack.

    Also, for repeat masking the idea would be to do 91, 1X by 1X block 
comparisons of your 91X data
set, and then compute repeat masks therefrom.  Of course if your data 
set is small then that doesn't
make sense (e.g. I don't bother for bacteria) but most folks are 
concentrating on large genomes.

    So hopefully the above gets you started in the right direction. I 
worked pretty hard to make things
modular and as simple as I could, but for HPC computes its hard to do 
and there is a bit to understand.
One way to get some deeper insight, would be to actually read through 
the dozen blog posts (and *not* the
command pages) at the Dazzler blog in reverse chronological order (see 
"reading list" below).
With each post, as I introduce the next set of commands or capabilities, 
I tried to give an intuitive
description of the concepts and commands.

    I am also happy to continue to answer specific questions as they occur.

    Best,  Gene

Blog Reading list:

https://dazzlerblog.wordpress.com/2014/03/22/the-dextractor-module-save-disk-space-for-your-pacbio-projects/
https://dazzlerblog.wordpress.com/2016/11/14/moving-to-the-sequel/
https://dazzlerblog.wordpress.com/2014/06/01/the-dazzler-db/
https://dazzlerblog.wordpress.com/2014/07/10/dalign-fast-and-sensitive-detection-of-all-pairwise-local-alignments/
https://dazzlerblog.wordpress.com/2014/12/31/damapper-and-other-dazzler-upgrades/
https://dazzlerblog.wordpress.com/2015/11/04/dazzler-data-and-results/
https://dazzlerblog.wordpress.com/2016/04/01/detecting-and-soft-masking-repeats/
https://dazzlerblog.wordpress.com/2015/11/05/trace-points/
https://dazzlerblog.wordpress.com/2016/05/21/dbs-and-dams-whats-the-difference/
https://dazzlerblog.wordpress.com/2016/05/29/seeing-your-reads/
https://dazzlerblog.wordpress.com/2016/07/31/damapper-mapping-your-reads/
https://dazzlerblog.wordpress.com/2015/11/06/intrinsic-quality-values/

**Comment by rrwick** [comment]
Thanks for the links - I'll read through those!

Regarding the split database, I was (at least for the time being) not bothering with splitting because my read sets are bacterial and max out at about 1 Gbp. I initially didn't bother with the DBsplit command, but I found that DASedit crashed if I didn't:

```
DASedit: reads.db is corrupted, read failed
```

If I included the DBsplit command, it didn't seem to affect anything else (e.g. the DAStrim output was identical), but it allowed DASedit to produce a patched read database. I assumed that DASedit only works on split databases. I'm not sure if this is a bug with DASedit or if it's intentional. I can share a dataset via Dropbox if you'd like help reproducing it - let me know.

Ryan
